### PR TITLE
Add Error handler

### PR DESCRIPTION
**Pull request by weskamm**
Currently sends email and chat message if failure is detected in validators.
If failure happen after validation, only chat messages are sent.
Depends on https://github.com/klips-project/rabbitmq-worker/pull/80/ to be merged and build as image.

can be tested via
```
docker-compose up --build
```
and then
```
cat workflows/publish-geotiff-with-validator-error.json | ~/Downloads/rabbitmqadmin -u rabbit -p rabbit publish exchange=amq.default routing_key=dispatcher
```

@hblitza @JakobMiksch please review

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -15,3 +15,15 @@ GEOSERVER_DEFAULT_USER=admin
...
+MAILHOST=my.mailserver.com
+MAILPORT=587
+MAILSECURE=true
+MAILAUTHUSER=user@mailserver.com
+MAILAUTHPASS=insertsecurepasswordhere
+MAILFROMSENDERNAME=Klips
+MAILFROMSENDEREMAIL=no-reply@mailserver.com
+
+# Mattermost
+MATTERMOST_HOOK_URL=https://chat.something.com/hooks/abcdefghi
```
Comment (JakobMiksch) [inline .env.example:29]: new line at the end of the file (says GitHub :sweat_smile: )

File: docker-compose.yml [hunk 2]
```diff
@@ -36,14 +39,60 @@ services:
...
+      - NODE_TLS_REJECT_UNAUTHORIZED=0
+
+  send-email:
+    image: ghcr.io/klips-project/mqm-worker/send-email:latest
+    environment:
+      - RABBITHOST=${RABBITMQ_HOSTNAME}
+      - RABBITUSER=${RABBITMQ_DEFAULT_USER}
+      - RABBITPASS=${RABBITMQ_DEFAULT_PASS}
+      - WORKERQUEUE=send-email
+      - RESULTSQUEUE=RABBITMQ_RESULTSQUEUE
```
Comment (JakobMiksch) [inline docker-compose.yml:84]: is this missing or am I confusing something?

```suggestion
      - RESULTSQUEUE=${RABBITMQ_RESULTSQUEUE}
```

File: worker/error-handler/index.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
+import amqp from 'amqplib';
```
Comment (JakobMiksch) [inline worker/error-handler/index.js:1]: a short comment what this worker does would be great

**Review by JakobMiksch — COMMENTED** [review]
Looking good so far. Some minor remarks inline so far.
I need to run it locally again.

File: worker/error-handler/index.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+          );
+          console.log('Email job dispatched from error handler');
+        }
+        const mattermostJob = {
+          "job": [
+            {
+              "id": 2,
+              "type": "send-mattermost-message",
+              "inputs": [
+                content + '```' + msg.content.toString() + '```\n1' +
```
[the remark this answers is not in the document]
Comment (weskamm) [reply worker/error-handler/index.js]: producing numbers ;-)

**Comment by weskamm** [comment]
fixed the mentioned issues. please check again

File: worker/error-handler/index.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+          recipients = job.email;
+        }
+
+        const subject = 'KLIPS - Job processing failed';
+        let content = `Processing of job with id ${job.id} failed.\n`;
+        if (failedTask) {
+          content += `The process of type '${failedTask} failed `;
+          content += `with the following error:\n\n`;
+        } else {
+          content += `The following error occured:\n\n`;
```
Comment (JakobMiksch) [inline worker/error-handler/index.js:58]:
```suggestion
          content += `The following error occurred:\n\n`;
```

File: worker/error-handler/index.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        const failedTask = job.nextTask?.task?.type;
+        let recipients;
+        if (failedTask && failedTask.toLowerCase().indexOf('validator') > -1) {
+          recipients = job.email;
+        }
+
+        const subject = 'KLIPS - Job processing failed';
+        let content = `Processing of job with id ${job.id} failed.\n`;
+        if (failedTask) {
+          content += `The process of type '${failedTask} failed `;
```
Comment (JakobMiksch) [inline worker/error-handler/index.js:55]:
```suggestion
          content += `The process of type '${failedTask}' failed `;
```

**Comment by weskamm** [comment]
@JakobMiksch: Example workflows updated to show where the email field should be provided

File: worker/error-handler/index.js [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+      try {
+        const job = JSON.parse(msg.content.toString());
+        console.log('Received a failed job in dead letter queue ...');
+
+        // Handling different recipients based on failed instance
+        // External failure: failures in validators -> mail to original job sender; POST message to chat
+        // Internal failure: failures after validators -> POST message to chat
+        const failedTask = job.nextTask?.task?.type;
+        let recipients;
+        if (failedTask && failedTask.toLowerCase().indexOf('validator') > -1) {
```
Comment (JakobMiksch) [inline worker/error-handler/index.js:48]: hm... takes me quite some time to understand this line. Could make a little explanation?

File: worker/error-handler/index.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    async function (msg) {
+      try {
+        const job = JSON.parse(msg.content.toString());
+        console.log('Received a failed job in dead letter queue ...');
+
+        // Handling different recipients based on failed instance
+        // External failure: failures in validators -> mail to original job sender; POST message to chat
+        // Internal failure: failures after validators -> POST message to chat
+        const failedTask = job.nextTask?.task?.type;
+        let recipients;
```
Comment (JakobMiksch) [inline worker/error-handler/index.js:47]: maybe rename to `emailRecipients` or so to indicate that these is only relevant for email

**Review by JakobMiksch — APPROVED** [review]
I tested it locally and it works. Great job. 
Only minor notes inside. Otherwise ready to merge.

**Comment by weskamm** [comment]
fixed the mentioned points. merging now

**Pull request merged**